File: frontend/src/api/llm.ts
import { api } from './client';

export interface NLPParseResult {
  title: string;
  description: string;
  estimated_minutes: number;
  priority: string;
  hard_deadline: string | null;
  scheduled_date: string | null;
  scheduled_time: string | null;
  category_names: string[];
}

export interface AgentActionResult {
  tool: string;
  args: Record<string, unknown>;
  result: Record<string, unknown>;
}

export interface AgentResponse {
  message: string;
  actions: AgentActionResult[];
}

export async function parseNLP(text: string): Promise<NLPParseResult> {
  const { data } = await api.post('/llm/parse-task', { text });
  return data;
}

export async function runAgent(text: string): Promise<AgentResponse> {
  const { data } = await api.post('/llm/agent', { text });
  return data;
}

export async function suggestPlan(date: string, pool_json: string, scheduled_json: string, special_json: string) {
  const { data } = await api.post('/llm/suggest-plan', {
    date, pool_tasks_json: pool_json, scheduled_tasks_json: scheduled_json, special_day_json: special_json,
  });
  return data;
}

export async function breakdownTask(goal: string, deadline: string | null) {
  const { data } = await api.post('/llm/breakdown', { goal, deadline });
  return data;
}

export async function smartSchedule(poolTaskIds: number[], start_date: string, end_date: string) {
  const { data } = await api.post('/llm/smart-schedule', {
    pool_task_ids: poolTaskIds, start_date, end_date,
  });
  return data;
}

export async function weeklySummary(start_date: string, end_date: string, tasks_json: string) {
  const { data } = await api.post('/llm/weekly-summary', {
    start_date, end_date, tasks_json,
  });
  return data;
}

File: backend/app/schemas/llm.py
"""Pydantic schemas for LLM endpoints."""

from typing import List, Optional
from datetime import date, time
from pydantic import BaseModel, Field


class NLPParseRequest(BaseModel):
    text: str


class NLPParseResponse(BaseModel):
    title: str
    description: str = ""
    estimated_minutes: int = 0
    priority: str = "P3"
    hard_deadline: Optional[date] = None
    scheduled_date: Optional[date] = None
    scheduled_time: Optional[time] = None
    category_names: List[str] = []


class WeeklySummaryRequest(BaseModel):
    start_date: date
    end_date: date
    tasks_json: str  # JSON string of task data for the week


class WeeklySummaryResponse(BaseModel):
    summary: str


class SmartScheduleRequest(BaseModel):
    pool_task_ids: List[int]
    start_date: date
    end_date: date


class SmartScheduleAssignment(BaseModel):
    task_id: int
    scheduled_date: date
    scheduled_time: Optional[time] = None
    reason: str = ""


class SmartScheduleResponse(BaseModel):
    assignments: List[SmartScheduleAssignment]


class BreakdownRequest(BaseModel):
    goal: str
    deadline: Optional[date] = None


class BreakdownSubTask(BaseModel):
    title: str
    description: str = ""
    estimated_minutes: int = 0
    priority: str = "P3"


class BreakdownResponse(BaseModel):
    sub_tasks: List[BreakdownSubTask]


class SuggestPlanRequest(BaseModel):
    date: date
    pool_tasks_json: str
    scheduled_tasks_json: str
    special_day_json: str = "{}"


class SuggestPlanResponse(BaseModel):
    plan: str
    suggestions: List[SmartScheduleAssignment]


class AnomalyCheckRequest(BaseModel):
    recent_history_json: str  # JSON string of recent task completion data


class AnomalyItem(BaseModel):
    type: str  # e.g., 'streak_break', 'low_completion', 'procrastination'
    message: str
    severity: str = "warning"  # 'info', 'warning', 'alert'


class AnomalyCheckResponse(BaseModel):
    anomalies: List[AnomalyItem]
